src/Conversion/ONNXToKrnl/RNN/LSTM.cpp

```diff
@@ -0,0 +1,605 @@
+//===--------------- LSTM.cpp - Lowering LSTM Op --------------------------===//
+//
+// Copyright 2019 The IBM Research Authors.
+//
+// =============================================================================
+//
+// This file lowers the ONNX LSTM Operators to Krnl dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/Conversion/ONNXToKrnl/RNN/RNNBase.hpp"
+
+using namespace mlir;
+
+struct LstmInputPack {
+  Value X;
+  Value W;
+  Value R;
+  Value Biofc;
+  Value sequenceLength;
+  Value initialHidden;
+  Value initialCell;
+  Value Piof;
+};
+
+struct LstmOutputPack {
+  Value Y;
+  Value Y_h;
+  Value Y_c;
+};
+
+struct LstmState {
+  Value allH;
+  Value ht;
+  Value ct;
+};
+
+struct LstmActivationPack {
+  RNNActivation f;
+  RNNActivation g;
+  RNNActivation h;
+};
+
+template <>
+std::tuple<LstmInputPack, LstmOutputPack>
+getInputOutputPack<ONNXLSTMOp, LstmInputPack, LstmOutputPack>(
+    Operation *op, ArrayRef<Value> operands) {
+  LstmInputPack inputPack;
+  inputPack.X = operands[0];
```

Comment:
Does the input pack add anything beyond what the operation adapter does?

Reply:
There is a subtle but important difference here is that `operands[0]` will return a MemRefType while `adapter.X()` will return a Tensor type.

Reply:
@tjingrant You are right. I used `ONNXLSTMOp rnnOp = llvm::dyn_cast<ONNXLSTMOp>(op);` and obtained X by `rnnOp.X()` which returned a Tensor type. Hence, I used `operands[0]` instead of `X`. If we use `ONNXLSTMOpOperandAdaptor`, it's OK. Will modify to use `ONNXLSTMOpOperandAdaptor`.